libvirt/resource_libvirt_domain.go

```diff
@@ -858,6 +877,39 @@ func resourceLibvirtDomainDelete(d *schema.ResourceData, meta interface{}) error
 		}
 	}
 
+	netIfacesCount := d.Get("network_interface.#").(int)
+	for i := 0; i < netIfacesCount; i++ {
+		prefix := fmt.Sprintf("network_interface.%d", i)
+		networkUUID, ok := d.GetOk(prefix + ".network_id")
+		if !ok {
+			continue
+		}
+		network, err := virConn.LookupNetworkByUUIDString(networkUUID.(string))
+		if err != nil {
+			return fmt.Errorf("Can't retrieve network ID %s", networkUUID)
+		}
+		defer network.Free()
+
+		networkName, err := network.GetName()
+		if err != nil {
+			return fmt.Errorf("Error retrieving network name: %s", err)
+		}
+		hostname := d.Get(prefix + ".hostname").(string)
+		mac := d.Get(prefix + ".mac").(string)
+		addresses := d.Get(prefix + ".addresses")
+		for _, addressI := range addresses.([]interface{}) {
+			address := addressI.(string)
+			ip := net.ParseIP(address)
+			if ip == nil {
+				return fmt.Errorf("Could not parse addresses '%s'", address)
+			}
+			log.Printf("[INFO] Removing IP/MAC/host=%s/%s/%s in '%s' network", ip.String(), mac, hostname, networkName)
+			if err := removeHost(&network, ip.String(), mac, hostname); err != nil {
+				return err
```

Comment:
This section of code is necessary in the case where a domain is destroyed but the network isn't.  I haven't thought of a way of figuring out if that "apply" has been triggered.